content/articles/optimizing-ml-models-with-optuna/index.md

```diff
@@ -0,0 +1,155 @@
+standardSc = StandardScaler()
+X = standardSc.fit_transform(X)
+# By dividing the dataset into train and test, the model is better able to reach a convergent state.
+from sklearn.model_selection import train_test_split
+X_train, X_test, y_train,y_test = train_test_split(X, y, test_size = 0.2, random_state = 12)
+```
+#### Testing the model
+It's easier to compare results when you've already built a baseline model. We can see how much better our model is after fine-tuning the parameters.
+```python
+import xgboost as gbst
+from sklearn.model_selection import cross_val_score
+xgboo_reg = gbst.XGBRegressor()
+# Cross-validation is a method for evaluating a score
+ourScores = cross_val_score(xgboo_reg, X_train,y_train , scoring = 'neg_root_mean_squared_error',
+                            #using all available CPUs
+                            n_jobs = -1,cv = 10)
+print(num.mean(ourScores), num.std(ourScores))
+print(ourScores)
+```
+Output:
+```bash
+-3.0784942308511307 0.42284534035667176
+[-2.62847993 -3.60582341 -3.31998107 -3.60355075 -3.03760446 -2.50243868 -2.92302011 -2.61990331 -3.67855489 -2.8655857 ]
+```
+#### Defining Functions
+`return_score()` provides a cross-verified score based on keyword arguments sent to it. We will take 1000 samples.
+```python
+def return_score(param):
+  ouNewModel = gbst.XGBRegressor(**param)  
+  regressionModelse = -num.mean(cross_val_score(ouNewModel,X_train[:1000],y_train[:10000], cv = 4, n_jobs =-1,scoring='neg_root_mean_squared_error'))
+  return regressionModelse
+```
+#### Importing optuna
+This is where we import Optuna.
+
+An objective function may be examined via a trial to see whether it successfully tunes.
+```python
+import optuna
+# The Samplers class defines our hyper-parameter space.
+from Optuna Import Trial, visualization
+from optuna.samplers import TPESampler
+```
+#### Creating the primary objective function
+We place all the possible hyper-parameter ranges, and the method accepts a trial object and should return a score value.
+```python
+def objective(trial):
+  parameter = {
+                "n_estimators" : trial.suggest_int('n_estimators', 0, 500),
+                'subsample':trial.suggest_uniform('subsample',0.4,0.9),
+                'gamma':trial.suggest_uniform('gamma', 0, 4),
+                'reg_alpha':trial.suggest_uniform('reg_alpha',0,6),
+               #uniformly distributed numbers between 0,2
+                'reg_lambda':trial.suggest_uniform('reg_lambda',0,2),
+                'min_child_weight':trial.suggest_int('min_child_weight',0,5),
+                'learning_rate':trial.suggest_loguniform('learning_rate',0.05,0.5),
+               #log distribution between [0.05,.5]
+                'colsample_bytree':trial.suggest_uniform('colsample_bytree',0.4,0.9),
+                'max_depth':trial.suggest_int('max_depth', 3, 5),
+                'nthread' : -1
+            }
+  #returns the regressionModelse score
+  return(return_score(parameter))
+```

```

Comment:
What are we doing here? Explain them in-detail. Why are doing this? How is it related?

Explain them as bulleted points.